app/features/worksheet_generator/tools.py

```diff
@@ -0,0 +1,481 @@
+        )
+
+        # Generate the final prompt by filling in the variables
+        filled_prompt = prompt.format(
+            grade_level=self.grade_level,
+            topic=self.topic,
+            difficulty_level=self.difficulty_level,
+            lang=self.lang
+        )
+
+        if self.verbose:
+            logger.info(f"Generated prompt: {filled_prompt}")
+
+        # Generate the question using the Google Gemini model
+        response = self.model.generate(prompts = [filled_prompt])
+
+        if self.verbose: logger.info("Response generation complete")
+    
+        # Extract the generated text from the response
+        output = response.generations[0][0].text.strip()
+
+        # Parse the JSON response using the configured parser
+        parsed_output = self.parser.parse(output)
+
+        if self.verbose: logger.info(f"Response parsing complete using {self.parser}")
+
+        return parsed_output
+    
+    @abstractmethod
+    def validate_response(self, response: Dict):
+        pass
+
+    @abstractmethod
+    def transform_json_dict(input_data: Dict) -> Dict:
+        pass
+
+    @abstractmethod
+    def create_questions(self, num_questions: int) -> List[Dict]:
+        pass
+
+class MultiChoiceQuestionBuilder(QuestionBuilder):
+    def __init__(self, grade_level, topic, difficulty_level, lang, prompt=None, model=None, parser=None, verbose=False):
+        default_config = {
+            "model": GoogleGenerativeAI(model="gemini-1.0-pro"),
+            "parser": JsonOutputParser(pydantic_object=MultiChoiceQuestion),
+            "prompt": read_text_file("prompt/multiple-choice-question-builder-prompt.txt")
+        }
+        
+        self.prompt = prompt or default_config["prompt"]
+        self.model = model or default_config["model"]
+        self.parser = parser or default_config["parser"]
+        
+        self.grade_level = grade_level
+        self.topic = topic
+        self.difficulty_level = difficulty_level
+        self.lang = lang
+        self.verbose = verbose
+        
+        if grade_level is None: raise ValueError("Grade level must be provided")
+        if topic is None: raise ValueError("Topic must be provided")
+        if difficulty_level is None: raise ValueError("Difficulty level must be provided")
+
+    def transform_json_dict(self, input_data: Dict) -> Dict:
+        # Validate and parse the input data to ensure it matches the MultiChoiceQuestion schema
+        quiz_question = MultiChoiceQuestion(**input_data)
+
+        # Transform the choices list into a dictionary
+        transformed_choices = {choice.key: choice.value for choice in quiz_question.choices}
+
+        # Create the transformed structure
+        transformed_data = {
+            "question": quiz_question.question,
+            "choices": transformed_choices,
+            "answer": quiz_question.answer,
+        }
+
+        return transformed_data
+    
+    def validate_response(self, response: Dict) -> bool:
```

Comment:
This validation is not required if the use of the JSONOutputParser is effectively done with the given prompt.